capsules/extra/src/app_loader.rs

```diff
@@ -0,0 +1,460 @@
+    ///            - The kernel fails to write a padding app (thereby potentially breaking the linkedlist)
+
+    fn command(
+        &self,
+        command_num: usize,
+        arg1: usize,
+        arg2: usize,
+        processid: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            0 => CommandReturn::success(),
+
+            1 => {
+                //setup phase
+                if self.current_user.is_none() {
+                    // No app is currently using the underlying storage.
+                    // Mark this app as active, and then execute the command.
+                    self.current_user.set(NonvolatileUser::App {
+                        processid: processid,
+                    });
+                }
+                let res = self.driver.setup(arg1); // pass the size of the app to the setup function
+                match res {
+                    Ok(app_len) => {
+                        self.new_app_length.set(app_len);
+                        CommandReturn::success()
+                    }
+
+                    Err(e) => {
+                        self.new_app_length.set(0);
+                        CommandReturn::failure(e)
+                    }
+                }
+            }
+
+            2 => {
+                // Request kernel to write app to flash
+
+                let res = self.enqueue_command(
+                    NonvolatileCommand::UserspaceWrite,
+                    arg1,
+                    arg2,
+                    Some(processid),
+                );
+                match res {
+                    Ok((offset, _len, pid)) => {
+                        let result = self
+                            .buffer
+                            .take()
+                            .map_or(Err(ErrorCode::RESERVE), |buffer| {
+                                let mut write_buffer = SubSliceMut::new(buffer);
+                                write_buffer.slice(..write_buffer.len());
+                                let res = self.driver.write_app_data(write_buffer, offset, pid);
+                                match res {
+                                    Ok(()) => Ok(()),
+                                    Err(e) => Err(e),
+                                }
+                            });
+                        match result {
+                            Ok(()) => CommandReturn::success(),
+                            Err(e) => {
+                                self.new_app_length.set(0);
+                                CommandReturn::failure(e)
+                            }
+                        }
+                    }
+                    Err(e) => {
+                        self.new_app_length.set(0);
+                        CommandReturn::failure(e)
+                    }
+                }
+            }
+
+            3 => {
+                // Request kernel to load the new app
+
+                let res = self.driver.load();
```

Comment:
Something here needs to check that the calling app is the current app.

Reply:
Yeah